Automation triggers examples (#11589)
PrefectHQ/prefect: docs/concepts/automations.md

Specify a name and, optionally, a description for the automation.

## Custom triggers

> Custom triggers allow advanced configuration of the conditions on which a trigger executes its actions. Several custom trigger fields accept values that end with trailing wildcards, like `"prefect.flow-run.*"`.

Reviewer: ```suggestion
Custom triggers allow advanced configuration of the conditions on which an automation executes its actions. Several custom trigger fields accept values that end with trailing wildcards, like `"prefect.flow-run.*"`.
```

![Viewing a custom trigger for automations for a workspace in Prefect Cloud.](/img/ui/automations-custom.png)

The schema that defines a trigger is as follows:

| Name               | Type               | Supports trailing wildcards | Description |
| ------------------ | ------------------ | --------------------------- | ----------- |

> | **match**          | object             | :material-check:            | Labels for resources which this Automation will match. |

Reviewer: nit: can all these `Name` fields be formatted as inline code? e.g. `match`?

Author: I had them that way, but mkdocs will wrap them onto multiple lines if they're code. I couldn't figure out how to fix that

| **match_related**  | object             | :material-check:            | Labels for related resources which this Automation will match. |

> | **after**          | array of strings   | :material-check:            | Event(s), one of which must have first been seen to start this automation. |

Reviewer: ```suggestion
| **after**          | set of strings   | :material-check:            | Event(s), one of which must have first been seen to start this automation. |
```

> | **expect**         | array of strings   | :material-check:            | The event(s) this automation is expecting to see. If empty, this automation will evaluate any matched event. |

Reviewer: ```suggestion
| **expect**         | set of strings   | :material-check:            | The event(s) this automation is expecting to see. If empty, this automation will evaluate any matched event. |
```

> | **for_each**       | array of strings   | :material-close:            | Evaluate the Automation separately for each distinct value of these labels on the resource. By default, labels refer to the primary resource of the triggering event. You may also refer to labels from related resources by specifying `related:<role>:<label>`. This will use the value of that label for the first related resource in that role. |

Reviewer: ```suggestion
| **for_each**       | set of strings   | :material-close:            | Evaluate the Automation separately for each distinct value of these labels on the resource. By default, labels refer to the primary resource of the triggering event. You may also refer to labels from related resources by specifying `related:<role>:<label>`. This will use the value of that label for the first related resource in that role. |
```

### Resource matching

`match` and `match_related` control which events a trigger considers for evaluation by filtering on the contents of their `resource` and `related` fields, respectively. Each label added to a `match` filter is `AND`ed with the other labels, and can accept a single value or a list of multiple values that are `OR`ed together.

Consider the `resource` and `related` fields on the following `prefect.flow-run.Completed` event, truncated for the sake of example. Its primary resource is a flow run, and since that flow run was started via a deployment, it is related to both its flow and its deployment:

```json
"resource": {
  "prefect.resource.id": "prefect.flow-run.925eacce-7fe5-4753-8f02-77f1511543db",
  "prefect.resource.name": "cute-kittiwake"
}
"related": [
  {
    "prefect.resource.id": "prefect.flow.cb6126db-d528-402f-b439-96637187a8ca",
    "prefect.resource.role": "flow",
    "prefect.resource.name": "hello"
  },
  {
    "prefect.resource.id": "prefect.deployment.37ca4a08-e2d9-4628-a310-cc15a323378e",
    "prefect.resource.role": "deployment",
    "prefect.resource.name": "example"
  }
]
```

> There are a number of valid ways to select the above event for evaluation, and the approach depends on the purpose of the automation.
> The following configuration will filter for any events whose primary resource is a flow run, _and_ that flow run has a name starting with `cute-` or `radical-`.

Reviewer: nit:
```suggestion
There are a number of valid ways to select the above event for evaluation, and the approach depends on the purpose of the automation.

The following configuration will filter for any events whose primary resource is a flow run, _and_ that flow run has a name starting with `cute-` or `radical-`.
```

### Expected events

Once an event has passed through the `match` filters, it must be decided if this event should be counted toward the trigger's `threshold`. Whether that is the case is determined by the event names present in `expect`.

This configuration informs the trigger to evaluate _only_ `prefect.flow-run.Completed` events that have passed the `match` filters.

```json
"expect": [
  "prefect.flow-run.Completed"
],
...
```

`threshold` decides the quantity of `expect`ed events needed to satisfy the trigger. Increasing the `threshold` above 1 will also require use of `within` to define a range of time in which multiple events are seen. The following configuration will expect two occurrences of `prefect.flow-run.Completed` within 60 seconds.

```json
"expect": [
  "prefect.flow-run.Completed"
],
"threshold": 2,
"within": 60,
...
```

`after` can be used to handle scenarios that require more complex event reactivity.

Take, for example, this flow which emits an event indicating the table it operates on is missing or empty:

```python
from prefect import flow
from prefect.events import emit_event
from db import Table

> @flow
> def transform(table_name: str):
>   table = Table(table_name)
> 
>   if not table.exists():
>     emit_event(
>         event="table-missing", resource={"prefect.resource.id": "etl-events.transform"}
>     )
>   elif table.is_empty():
>     emit_event(
>         event="table-empty", resource={"prefect.resource.id": "etl-events.transform"}
>     )
>   else:
>     # transform data

Reviewer: nit: easier to read
```suggestion
@flow
def transform(table_name: str):
  table = Table(table_name)

  if not table.exists():
    emit_event(
        event="table-missing",
        resource={"prefect.resource.id": "etl-events.transform"}
    )
  elif table.is_empty():
    emit_event(
        event="table-empty",
        resource={"prefect.resource.id": "etl-events.transform"}
    )
  else:
    # transform data
```

```

> The following configuration uses `after` to prevent this automation from firing unless either a `table-missing` or a `table-empty` event has occurred before a flow run of this deployment completes. Note how `match` and `match_related` are used to ensure the trigger only evaluates events that are relevant to its purpose:

Reviewer: nit: perhaps this sentence

> Note how `match` and `match_related` are used to ensure the trigger only evaluates events that are relevant to its purpose:

could be moved to a `!!! Tip` below the json snippet so its prettier?
